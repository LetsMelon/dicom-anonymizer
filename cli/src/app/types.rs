use anyhow::Result;
use clap::{Arg, ArgMatches, Command};
use yaml_rust::Yaml;

pub type StaticCommand = Command<'static>;
pub type StaticArg = Arg<'static>;

pub trait IMatcher<T> {
    fn match_args(matches: ArgMatches) -> Result<Box<Self>>;
    fn match_trait(&self) -> Result<T>;
}

pub trait IConfigFile {
    fn parse(content: Vec<Yaml>) -> Result<Box<Self>>;
    fn get_version() -> String;
}
