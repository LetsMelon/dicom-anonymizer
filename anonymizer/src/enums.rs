use std::str::FromStr;
use derive_more::{Display};
use dicom_core::Tag;
use tags_list::List as TagsList;
use strum::EnumCount;

#[derive(Display, Clone, Debug, EnumCount)]
pub enum PatientSex {
    M,
    F,
    O,
}

impl FromStr for PatientSex {
    type Err = ::strum::ParseError;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        assert!(PatientSex::COUNT == 3, "TODO: implement new patient sex value");
        match s.to_ascii_lowercase().as_str() {
            "m" => Ok(PatientSex::M),
            "f" => Ok(PatientSex::F),
            "o" => Ok(PatientSex::O),
            _ => Err(::strum::ParseError::VariantNotFound)
        }
    }
}

impl PatientSex {
    pub fn value(&self) -> &str {
        match *self {
            PatientSex::M => "M",
            PatientSex::F => "F",
            PatientSex::O => "O",
        }
    }
}

impl Default for PatientSex {
    fn default() -> Self { PatientSex::M }
}

pub enum RemoveTagsInput {
    Vec(Vec<Tag>),
    List(TagsList),
    VecList(Vec<TagsList>)
}

impl From<Vec<Tag>> for RemoveTagsInput {
    fn from(v: Vec<Tag>) -> Self {
        RemoveTagsInput::Vec(v)
    }
}

impl From<TagsList> for RemoveTagsInput {
    fn from(t: TagsList) -> Self {
        RemoveTagsInput::List(t)
    }
}

impl From<Vec<TagsList>> for RemoveTagsInput {
    fn from(vt: Vec<TagsList>) -> Self {
        RemoveTagsInput::VecList(vt)
    }
}
