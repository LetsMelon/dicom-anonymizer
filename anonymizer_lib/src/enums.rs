use derive_more::Display;
use dicom_core::Tag;
use serde::{Deserialize, Serialize};
use std::str::FromStr;
use strum::EnumCount;
use tags_list_lib::List as TagsList;

use crate::tag::CustomTag;

#[derive(Display, Copy, Clone, Debug, EnumCount, Serialize, Deserialize, Eq, PartialEq)]
pub enum PatientSex {
    M,
    F,
    O,
}

impl FromStr for PatientSex {
    type Err = ::strum::ParseError;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        assert!(
            PatientSex::COUNT == 3,
            "TODO: implement new patient sex value"
        );
        match s.to_ascii_lowercase().as_str() {
            "m" => Ok(PatientSex::M),
            "f" => Ok(PatientSex::F),
            "o" => Ok(PatientSex::O),
            _ => Err(::strum::ParseError::VariantNotFound),
        }
    }
}

impl PatientSex {
    pub fn value(&self) -> &str {
        match *self {
            PatientSex::M => "M",
            PatientSex::F => "F",
            PatientSex::O => "O",
        }
    }
}

impl Default for PatientSex {
    fn default() -> Self {
        PatientSex::M
    }
}

pub enum RemoveTagsInput {
    Vec(Vec<CustomTag>),
    List(TagsList),
    VecList(Vec<TagsList>),
}

impl From<Vec<Tag>> for RemoveTagsInput {
    fn from(v_t: Vec<Tag>) -> Self {
        RemoveTagsInput::Vec(CustomTag::from_vec(v_t))
    }
}

impl From<Vec<CustomTag>> for RemoveTagsInput {
    fn from(v_ct: Vec<CustomTag>) -> Self {
        RemoveTagsInput::Vec(v_ct)
    }
}

impl From<TagsList> for RemoveTagsInput {
    fn from(t: TagsList) -> Self {
        RemoveTagsInput::List(t)
    }
}

impl From<Vec<TagsList>> for RemoveTagsInput {
    fn from(vt: Vec<TagsList>) -> Self {
        RemoveTagsInput::VecList(vt)
    }
}

#[derive(Clone, Debug, Eq, PartialEq, Serialize, Deserialize)]
pub enum TagAction<T> {
    /// Change the tag with the given value, similar to `Option::Some(T)`
    Change(T),
    /// Changes nothing, similar to `Option::None`
    Keep,
    /// Remove the tag
    Remove,
}

impl<T> TagAction<T> {
    /// Maps an `TagAction::Change<T>` to `Option<U>` by applying a function to a contained value.
    /// Copied from https://doc.rust-lang.org/std/option/enum.Option.html#method.map
    #[inline]
    pub fn map<U, F>(self, f: F) -> TagAction<U>
    where
        F: FnOnce(T) -> U,
    {
        match self {
            TagAction::Change(v) => TagAction::Change(f(v)),
            TagAction::Keep => TagAction::Keep,
            TagAction::Remove => TagAction::Remove,
        }
    }
}

impl<T> Default for TagAction<T> {
    fn default() -> Self {
        TagAction::Keep
    }
}

impl<T> Into<Option<T>> for TagAction<T> {
    fn into(self) -> Option<T> {
        match self {
            TagAction::Change(value) => Some(value),
            _ => None,
        }
    }
}

impl<T> Into<TagAction<T>> for Option<T> {
    fn into(self) -> TagAction<T> {
        match self {
            None => TagAction::Keep,
            Some(value) => TagAction::Change(value),
        }
    }
}
