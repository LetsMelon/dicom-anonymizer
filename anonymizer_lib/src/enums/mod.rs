mod patient_sex;
mod remove_tags_input;
mod tag_action;

pub use patient_sex::PatientSex;
pub use remove_tags_input::RemoveTagsInput;
pub use tag_action::TagAction;
