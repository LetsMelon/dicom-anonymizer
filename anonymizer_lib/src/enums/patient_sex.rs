use derive_more::Display;
use serde::{Deserialize, Serialize};
use std::str::FromStr;
use strum::EnumCount;

#[derive(Display, Copy, Clone, Debug, EnumCount, Serialize, Deserialize, Eq, PartialEq)]
pub enum PatientSex {
    M,
    F,
    O,
}

impl FromStr for PatientSex {
    type Err = ::strum::ParseError;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        assert!(
            PatientSex::COUNT == 3,
            "TODO: implement new patient sex value"
        );
        match s.to_ascii_lowercase().as_str() {
            "m" => Ok(PatientSex::M),
            "f" => Ok(PatientSex::F),
            "o" => Ok(PatientSex::O),
            _ => Err(::strum::ParseError::VariantNotFound),
        }
    }
}

impl PatientSex {
    pub fn value(&self) -> &str {
        match *self {
            PatientSex::M => "M",
            PatientSex::F => "F",
            PatientSex::O => "O",
        }
    }
}

impl Default for PatientSex {
    fn default() -> Self {
        PatientSex::M
    }
}

#[cfg(test)]
mod tests {
    use super::PatientSex;
    use strum::EnumCount;

    mod from_str {
        use super::PatientSex;
        use std::str::FromStr;

        #[test]
        fn ok() {
            assert_eq!(PatientSex::M, PatientSex::from_str("M").unwrap());
            assert_eq!(PatientSex::F, PatientSex::from_str("F").unwrap());
            assert_eq!(PatientSex::O, PatientSex::from_str("O").unwrap());
            assert_eq!(PatientSex::M, PatientSex::from_str("m").unwrap());
            assert_eq!(PatientSex::F, PatientSex::from_str("f").unwrap());
            assert_eq!(PatientSex::O, PatientSex::from_str("o").unwrap());
        }

        #[test]
        fn error() {
            let ps = PatientSex::from_str("MM");

            match ps {
                Ok(_) => assert!(false, "Should throw an error"),
                Err(_) => assert!(true),
            };
        }
    }

    #[test]
    fn enum_count() {
        assert_eq!(PatientSex::COUNT, 3);
    }
}
