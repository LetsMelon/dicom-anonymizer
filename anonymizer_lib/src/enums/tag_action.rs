use serde::{Deserialize, Serialize};

#[derive(Clone, Debug, Eq, PartialEq, Serialize, Deserialize)]
pub enum TagAction<T> {
    /// Change the tag with the given value, similar to `Option::Some(T)`
    Change(T),
    /// Changes nothing, similar to `Option::None`
    Keep,
    /// Remove the tag
    Remove,
}

impl<T> TagAction<T> {
    /// Maps an `TagAction::Change<T>` to `Option<U>` by applying a function to a contained value.
    /// Copied from https://doc.rust-lang.org/std/option/enum.Option.html#method.map
    #[inline]
    pub fn map<U, F>(self, f: F) -> TagAction<U>
    where
        F: FnOnce(T) -> U,
    {
        match self {
            TagAction::Change(v) => TagAction::Change(f(v)),
            TagAction::Keep => TagAction::Keep,
            TagAction::Remove => TagAction::Remove,
        }
    }
}

impl<T> Default for TagAction<T> {
    fn default() -> Self {
        TagAction::Keep
    }
}

impl<T> From<Option<T>> for TagAction<T> {
    fn from(opt: Option<T>) -> Self {
        match opt {
            None => TagAction::default(),
            Some(value) => TagAction::Change(value),
        }
    }
}

impl<T> From<TagAction<T>> for Option<T> {
    fn from(tag_action: TagAction<T>) -> Self {
        match tag_action {
            TagAction::Change(value) => Option::Some(value),
            TagAction::Keep => Option::None,
            TagAction::Remove => Option::None,
        }
    }
}

#[cfg(test)]
mod tests {
    use super::TagAction;

    #[test]
    fn has_option_map_like_function() {
        assert_eq!(
            TagAction::Change("MyString"),
            TagAction::Change(1).map(|_| "MyString")
        );
        assert_eq!(
            TagAction::Change("MyString"),
            TagAction::Change(Option::Some("MyString")).map(|value| value.unwrap())
        );
        assert_ne!(
            TagAction::Change("MyString"),
            TagAction::Keep.map(|_: String| "MyString")
        );
    }

    #[test]
    fn can_be_transformed_into_an_option_and_back() {
        type T = usize;

        assert_eq!(
            Option::Some("MyString"),
            Option::from(TagAction::Change("MyString")),
        );
        assert_eq!(Option::<T>::None, Option::from(TagAction::<T>::Keep),);
        assert_eq!(Option::<T>::None, Option::from(TagAction::<T>::Remove),);

        assert_eq!(TagAction::Change(1), TagAction::from(Option::Some(1)),);
        assert_eq!(TagAction::<T>::Keep, TagAction::from(Option::<T>::None),);
    }
}
