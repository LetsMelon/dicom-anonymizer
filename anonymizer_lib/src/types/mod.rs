mod dicom_date_time;
mod tag;

pub use dicom_date_time::CustomDicomDateTime;
pub use tag::CustomTag;
