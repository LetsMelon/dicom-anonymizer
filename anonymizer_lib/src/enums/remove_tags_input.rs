use dicom_core::Tag;
use tags_list_lib::List as TagsList;

use crate::types::CustomTag;

pub enum RemoveTagsInput {
    Vec(Vec<CustomTag>),
    List(TagsList),
    VecList(Vec<TagsList>),
}

impl From<Vec<Tag>> for RemoveTagsInput {
    fn from(v_t: Vec<Tag>) -> Self {
        RemoveTagsInput::Vec(CustomTag::from_vec(v_t))
    }
}

impl From<Vec<CustomTag>> for RemoveTagsInput {
    fn from(v_ct: Vec<CustomTag>) -> Self {
        RemoveTagsInput::Vec(v_ct)
    }
}

impl From<TagsList> for RemoveTagsInput {
    fn from(t: TagsList) -> Self {
        RemoveTagsInput::List(t)
    }
}

impl From<Vec<TagsList>> for RemoveTagsInput {
    fn from(vt: Vec<TagsList>) -> Self {
        RemoveTagsInput::VecList(vt)
    }
}
