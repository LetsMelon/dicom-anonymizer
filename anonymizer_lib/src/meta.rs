use serde::{Deserialize, Serialize};

use crate::enums::{PatientSex, RemoveTagsInput};
use crate::types::{CustomDicomDateTime, CustomTag};
use crate::TagAction;

#[derive(Debug, Builder, Clone, Serialize, Deserialize, Eq, PartialEq)]
#[builder(derive(Debug))]
pub struct AnonymizerMeta {
    #[builder(setter(into, strip_option), default)]
    pub(crate) patient_name: TagAction<String>,

    #[builder(setter(into, strip_option), default)]
    pub(crate) patient_birth_date: TagAction<CustomDicomDateTime>,

    #[builder(setter(custom, into, strip_option), default)]
    pub(crate) remove_tags: Vec<CustomTag>,

    #[builder(setter(into, strip_option), default)]
    pub(crate) patient_sex: TagAction<PatientSex>,
}

impl AnonymizerMetaBuilder {
    pub fn remove_tag(&mut self, value: CustomTag) -> &mut Self {
        let mut obj = self;

        if obj.remove_tags.is_none() {
            obj.remove_tags = Some(Vec::<CustomTag>::new());
        }
        obj.remove_tags.as_mut().unwrap().push(value);

        obj
    }

    pub fn remove_tags(&mut self, values: RemoveTagsInput) -> &mut Self {
        let obj = self;

        match values {
            RemoveTagsInput::Vec(v) => {
                for item in v {
                    obj.remove_tag(item);
                }
            }
            RemoveTagsInput::List(t) => {
                obj.remove_tags(RemoveTagsInput::from(CustomTag::from_vec(t.value())));
            }
            RemoveTagsInput::VecList(vt) => {
                for item in vt {
                    obj.remove_tags(RemoveTagsInput::from(item));
                }
            }
        }

        obj
    }
}

#[cfg(test)]
mod tests {

    mod serialize {
        use crate::types::{CustomDicomDateTime, CustomTag};
        use crate::{AnonymizerMeta, PatientSex, TagAction};
        use chrono::FixedOffset;
        use dicom_core::value::{DicomDate, DicomDateTime};

        #[test]
        fn default() {
            let am = AnonymizerMeta {
                patient_name: TagAction::default(),
                patient_birth_date: TagAction::default(),
                remove_tags: vec![],
                patient_sex: TagAction::default(),
            };
            insta::assert_json_snapshot!(am);
        }

        #[test]
        fn basic_values() {
            let am = AnonymizerMeta {
                patient_name: TagAction::Remove,
                patient_birth_date: TagAction::Keep,
                remove_tags: vec![CustomTag {
                    group: 0,
                    element: 0,
                }],
                patient_sex: TagAction::Remove,
            };
            insta::assert_json_snapshot!(am);
        }

        #[test]
        fn complex_values() {
            let am = AnonymizerMeta {
                patient_name: TagAction::Change("New Patient Name".to_string()),
                patient_birth_date: TagAction::Change(CustomDicomDateTime::new(
                    DicomDateTime::from_date(
                        DicomDate::from_ymd(2016, 8, 12).unwrap(),
                        FixedOffset::east(0),
                    ),
                )),
                remove_tags: vec![CustomTag {
                    group: 0,
                    element: 0,
                }],
                patient_sex: TagAction::Change(PatientSex::O),
            };
            insta::assert_json_snapshot!(am);
        }
    }
}
