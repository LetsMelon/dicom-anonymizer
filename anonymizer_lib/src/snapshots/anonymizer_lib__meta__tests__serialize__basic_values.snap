---
source: anonymizer_lib/src/meta.rs
expression: am
---
{
  "patient_name": "Remove",
  "patient_birth_date": "Keep",
  "remove_tags": [
    {
      "group": 0,
      "element": 0
    }
  ],
  "patient_sex": "Remove"
}
